### PR TITLE
Modularise training widget inputs

### PR DESCRIPTION
**Pull request by alessandrofelder**
This PR
* separates the (abstract) `InputContainer` class into its own file
* adds (concrete) data classes to represent input sections of the training widget (equivalent design to the detection widget)
* adds a smoke test for the training widget.

I've done some manual testing to check that the correct cellfinder core functionality is called, but not more.

Partially addresseds #41 - still need to tackle the curation widget.

### CODE REVIEW

File: cellfinder_napari/train/train_containers.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+@dataclass
+class OptionalTrainingInputs(InputContainer):
+    continue_training: bool = False
+    augment: bool = True
+    tensorboard: bool = False
+    save_weights: bool = False
+    save_checkpoints: bool = True
+    save_progress: bool = True
+    epochs: int = 100
+    learning_rate: float = 0.0001
```
Comment (paddyroddy) [inline cellfinder_napari/train/train_containers.py:79]:
```suggestion
    learning_rate: float = 1e-4
```

**Review by paddyroddy — CHANGES_REQUESTED** [review]
Quite a lot of questions/suggestions. Nothing major

File: cellfinder_napari/input_container.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    @abstractmethod
+    def as_core_arguments(self) -> dict:
+        """Determines how dataclass fields are passed to cellfinder-core.
+
+        The implementation provided here can be re-used in derived classes, if convenient.
+        """
+        # note that asdict returns a new instance of a dict,
+        # so any subsequent modifications of this dict won't affect the class instance
+        return asdict(self)
```
[the remark this answers is not in the document]
Comment (alessandrofelder) [reply cellfinder_napari/input_container.py:34]: Then this is not used, unless you want to, in which case you should call `super().as_core_arguments`?

File: cellfinder_napari/train/train_containers.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        )
+
+
+@dataclass
+class OptionalNetworkInputs(InputContainer):
+    """Container for Optional Network input widgets"""
+
+    trained_model: Optional[Path] = Path.home()
+    model_weights: Optional[Path] = Path.home()
+    model_depth: str = "50"
```
[the remark this answers is not in the document]
Comment (alessandrofelder) [reply cellfinder_napari/train/train_containers.py:45]: It's what the backend expects...

File: cellfinder_napari/train/train_containers.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    learning_rate: float = 0.0001
+    batch_size: int = 16
+    test_fraction: float = 0.1
+
+    def as_core_arguments(self) -> dict:
+        arguments = super().as_core_arguments()
+        arguments["no_augment"] = not arguments.pop("augment")
+        arguments["no_save_checkpoints"] = not arguments.pop(
+            "save_checkpoints"
+        )
```
[the remark this answers is not in the document]
Comment (alessandrofelder) [reply cellfinder_napari/train/train_containers.py:88]: Yes, the [argument of `pop` is a dictionary key](https://docs.python.org/3/library/stdtypes.html#dict.pop), in this case a string. The return value is a bool, so the `not` negates that bool.

I thought this was a good succinct way of removing the key "augment" with a "no_augment" key (with the opposite value) that the backed expects...

File: cellfinder_napari/train/train_containers.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        return arguments
+
+    @classmethod
+    def widget_representation(cls) -> dict:
+        return dict(
+            data_options=html_label_widget("Training Data:"),
+            yaml_files=cls._custom_widget(
+                "yaml_files",
+                custom_label="YAML files",
+                mode="rm",
```
[the remark this answers is not in the document]
Comment (alessandrofelder) [reply cellfinder_napari/train/train_containers.py:32]: Yes, these are options that are passed down to `napari`'s widget creation functionality, in this case a [`magicgui FileEdit`](https://napari.org/magicgui/usage/_autosummary/magicgui.widgets.FileEdit.html#magicgui.widgets.FileEdit) widget (It decides it's a FileEdit because we pass a `Path` as the value).

Looks like the ["rm" stands for "**r**eturn one or **m**ore existing files" ](https://napari.org/magicgui/api/types.html#magicgui.types.FileDialogMode) which we need here as we may want to specify a number of YAML files to be used.

File: cellfinder_napari/detect/detect_containers.py [hunk 6]
```diff
@@ -6,72 +6,10 @@
 import napari
 import numpy
 
+from cellfinder_napari.input_container import InputContainer
```
[the remark this answers is not in the document]
Comment (paddyroddy) [reply cellfinder_napari/detect/detect_containers.py:9]: Okay, happy if you're happy. Just made me hesitate

File: cellfinder_napari/input_container.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            else custom_label
+        )
+        value = cls.defaults()[key]
+        return dict(value=value, label=label, **kwargs)
+
+    @classmethod
+    @abstractmethod
+    def widget_representation(cls) -> dict:
+        """What the class will look like as a napari widget"""
+        pass
```
[the remark this answers is not in the document]
Comment (paddyroddy) [reply cellfinder_napari/input_container.py:57]: Discussed this with @dstansby recently. This isn't actually required and isn't executed anyway. Personally I like this as I always view `pass` as a placeholder for a method that needs to be filled.

[the remark this answers is not in the document]
Comment (dstansby) [reply cellfinder_napari/input_container.py:57]: Isn't this exactly the situation where this (ie. the current `pass` statement) is a placeholder for a method that needs to be filled?

[the remark this answers is not in the document]
Comment (paddyroddy) [reply cellfinder_napari/input_container.py:57]: Well, yeah, but in the child class not the parent

File: cellfinder_napari/train/train_containers.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (alessandrofelder) [reply cellfinder_napari/train/train_containers.py:45]: I've made this clearer as to what this actually means. It's the first key of a set of keys that each point to an existing model of that depth in the backend IIUC.

File: cellfinder_napari/train/train.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        augment: bool,
+        tensorboard: bool,
+        save_weights: bool,
+        save_checkpoints: bool,
+        save_progress: bool,
+        epochs: int,
+        learning_rate: float,
+        batch_size: int,
+        test_fraction: float,
+        misc_options,
```
[the remark this answers is not in the document]
Comment (alessandrofelder) [reply cellfinder_napari/train/train.py:58]: It's a dict (specifically one that napari then interprets to make an html label)

File: cellfinder_napari/train/train.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+            Training batch size
+        test_fraction : float
+            Fraction of training data to use for validation
+        number_of_free_cpus : int
+            How many CPU cores to leave free
+        reset_button :
+            Reset parameters to default
+        """
+        if yaml_files[0] == Path.home():  # type: ignore
+            print("Please select a YAML file for training")
```
[the remark this answers is not in the document]
Comment (alessandrofelder) [reply cellfinder_napari/train/train.py:110]: Indeed - several things could be improved.
What I've addressed in response to this review comment is to use `show_info` from `napari.utils.notifications` so the problem is shown to the user more directly.

As for the `Path.home()` meaning use the default in some cases, or equating to a non-yet-specified, required input as in this case, I've extended the description #56.

File: cellfinder_napari/input_container.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (paddyroddy) [reply cellfinder_napari/input_container.py:57]: I think when I first showed `abstractmethod` this was how I was shown to do it, so it's my bias

File: cellfinder_napari/train/train_containers.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (alessandrofelder) [reply cellfinder_napari/train/train_containers.py:32]: Good point. I now use the `magicgui.types` enum `FileDialogMode.EXISTING_FILES` directly which should make this clearer.

File: cellfinder_napari/train/train.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+    OptionalNetworkInputs,
+    OptionalTrainingInputs,
+    TrainingDataInputs,
+)
+
+
+def train() -> FunctionGui:
+    @magicgui(
+        header=header_label_widget,
+        training_label=html_label_widget("Network training", "h3"),
```
Comment (paddyroddy) [inline cellfinder_napari/train/train.py:29]: I think we should use the `*` operator in the `html_label_widget` function definition such that `tag` is a keyword argument
```suggestion
        training_label=html_label_widget("Network training", tag="h3"),
```

File: cellfinder_napari/train/train.py [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+        **MiscTrainingInputs.widget_representation(),
+        call_button=True,
+        reset_button=dict(widget_type="PushButton", text="Reset defaults"),
+    )
+    def widget(
+        header,
+        training_label,
+        data_options: dict,
+        yaml_files: Path,
+        output_directory: Path,
```
[the remark this answers is not in the document]
Comment (paddyroddy) [reply cellfinder_napari/train/train.py:42]: Thoughts on this @alessandrofelder ?

File: cellfinder_napari/train/train.py [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+        epochs: int,
+        learning_rate: float,
+        batch_size: int,
+        test_fraction: float,
+        misc_options: dict,
+        number_of_free_cpus: int,
+        reset_button: PushButton,
+    ):
+        """
+
```
Comment (paddyroddy) [inline cellfinder_napari/train/train.py:64]: Any reason for this gap?
```suggestion
```

File: cellfinder_napari/train/train.py [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+        batch_size : int
+            Training batch size
+        test_fraction : float
+            Fraction of training data to use for validation
+        number_of_free_cpus : int
+            How many CPU cores to leave free
+        reset_button : PushButton
+            Reset parameters to default
+        """
+
```
Comment (paddyroddy) [inline cellfinder_napari/train/train.py:110]: Same here?
```suggestion
```

File: cellfinder_napari/train/train.py [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+
+@thread_worker
+def run_training(
+    training_data_inputs: TrainingDataInputs,
+    optional_network_inputs: OptionalNetworkInputs,
+    optional_training_inputs: OptionalTrainingInputs,
+    misc_training_inputs: MiscTrainingInputs,
+):
+    from cellfinder_core.train.train_yml import run as train_yml
```
Comment (paddyroddy) [inline cellfinder_napari/train/train.py:175]: Why is this being imported here and not at the top?

File: cellfinder_napari/train/train_containers.py [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+        return dict(
+            training_options=html_label_widget("Training (optional)"),
+            continue_training=cls._custom_widget("continue_training"),
+            augment=cls._custom_widget("augment"),
+            tensorboard=cls._custom_widget("tensorboard"),
+            save_weights=cls._custom_widget("save_weights"),
+            save_checkpoints=cls._custom_widget("save_checkpoints"),
+            save_progress=cls._custom_widget("save_progress"),
+            epochs=cls._custom_widget("epochs"),
+            learning_rate=cls._custom_widget("learning_rate", step=0.0001),
```
Comment (paddyroddy) [inline cellfinder_napari/train/train_containers.py:105]:
```suggestion
            learning_rate=cls._custom_widget("learning_rate", step=1e-4),
```

**Review by paddyroddy — APPROVED** [review]

File: cellfinder_napari/train/train.py (hunk 14 above)
Comment (alessandrofelder) [reply cellfinder_napari/train/train.py:173]: Good point. Moved.

File: cellfinder_napari/train/train.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (alessandrofelder) [reply cellfinder_napari/train/train.py:42]: Napari crashes ~silently when I change this to `os.PathLike` so I will avoid it for now at least.
This is presumably because magicgui doesn't know what widget is should create for an `os.PathLike` object, but it does for `Path`?

[the remark this answers is not in the document]
Comment (paddyroddy) [reply cellfinder_napari/train/train.py:43]: @dstansby I agree when we know that should be strict. I wasn't sure if we knew here or not
